ksql-engine/src/test/java/io/confluent/ksql/KsqlEngineTest.java

```diff
@@ -535,6 +537,91 @@ public void shouldCleanUpInternalTopicSchemasFromSchemaRegistry() throws Excepti
     )));
   }
 
+  @Test
+  public void shouldCleanUpInternalTopicsOnClose() {
+    // Given:
+    final QueryMetadata query = ksqlEngine.execute(
+        "select * from test1;",
+        KSQL_CONFIG, Collections.emptyMap()).get(0);
+
+    query.start();
+
+    // When:
+    query.close();
+
+    // Then:
+    verify(topicClient).deleteInternalTopics(query.getQueryApplicationId());
+  }
+
+  @Test
+  public void shouldNotCleanUpInternalTopicsOnCloseIfQueryNeverStarted() {
+    // Given:
+    final QueryMetadata query = ksqlEngine.execute(
+        "create stream s1 with (value_format = 'avro') as select * from test1;",
+        KSQL_CONFIG, Collections.emptyMap()).get(0);
+
+    // When:
+    query.close();
+
+    // Then:
+    verify(topicClient, never()).deleteInternalTopics(any());
+  }
+
+  @Test
+  public void shouldRemovePersistentQueryFromEngineWhenTerminated() {
+    // Given:
+    final long startingLiveQueries = ksqlEngine.numberOfLiveQueries();
+    final long startingPersistentQueries = ksqlEngine.numberOfPersistentQueries();
+
+    final QueryMetadata query = ksqlEngine.execute(
+        "create stream s1 with (value_format = 'avro') as select * from test1;",
+        KSQL_CONFIG, Collections.emptyMap()).get(0);
+
+    // When:
+    ksqlEngine.terminateQuery(getQueryId(query));
+
+    // Then:
+    assertThat(ksqlEngine.getPersistentQuery(getQueryId(query)), is(nullValue()));
+    assertThat(ksqlEngine.numberOfLiveQueries(), is(startingLiveQueries));
+    assertThat(ksqlEngine.numberOfPersistentQueries(), is(startingPersistentQueries));
+  }
+
+  @Test
+  public void shouldRemovePersistentQueryFromEngineWhenClosed() {
+    // Given:
+    final long startingLiveQueries = ksqlEngine.numberOfLiveQueries();
+    final long startingPersistentQueries = ksqlEngine.numberOfPersistentQueries();
+
+    final QueryMetadata query = ksqlEngine.execute(
+        "create stream s1 with (value_format = 'avro') as select * from test1;",
+        KSQL_CONFIG, Collections.emptyMap()).get(0);
+
+
+    // When:
+    query.close();
```

Comment:
Same here, should we test this in `QueryMetadataTest`? or maybe move this to `// Given:` section.

Reply:
See above.